core/state/statedb.go

```diff
@@ -407,11 +430,13 @@ func (s *StateDB) CommitBatch() (root common.Hash, batch ethdb.Batch) {
 func (s *StateDB) commit(dbw trie.DatabaseWriter) (common.Hash, error) {
 	s.refund = new(big.Int)
 
-	for _, stateObject := range s.stateObjects {
+	// Commit objects to the trie.
+	for addr, stateObject := range s.stateObjects {
 		if stateObject.remove {
 			// If the object has been removed, don't bother syncing it
 			// and just mark it for deletion in the trie.
 			s.DeleteStateObject(stateObject)
+			delete(s.all, addr)
```

Comment:
The cache is retained as long as blocks follow a progression. It gets obliterated only if a new block is being imported that has a different parent than the previous import (since then the cached state is invalid from the new block's perspective).

This could be worked around with a journal for the cache, but that's a whole new can of worms.